### PR TITLE
Convert a8, aO and ao command to rzshell

### PR DESCRIPTION
**Pull request by PeiweiHu**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This pr ports command ao, aO, and a8 to new rzshell. The following changes are performed in detail.

(e->esil, s->size, d->description)

+ split function core_analysis_bytes() to core_analysis_bytes_esil/size/desc/json/standard()
+ aO(j), aOd, aOs, aOe in rzshell
+ a8(j) in rzshell
+ ao(j), aos, aoe, aod, aoda, aoc, aom, aoma in rzshell (Note that aoef and ao* don't have implementation in old shell. Thus they are removed. )

...

**Test plan**

CI is green. 

**Closing issues**

partially solve https://github.com/rizinorg/rizin/issues/1442

...

### CODE REVIEW

File: librz/core/cmd_descs/cmd_analysis.yaml [hunk 1]
```diff
@@ -1453,3 +1453,13 @@ commands:
               - name: class_name
                 type: RZ_CMD_ARG_TYPE_STRING
                 optional: true
+  - name: a8
+    summary: Analyze bytes
+    cname: analyze_bytes
+    args:
+      - name: hexpairs
+        type: RZ_CMD_ARG_TYPE_STRING
+    type: RZ_CMD_DESC_TYPE_ARGV_MODES
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml]:
```suggestion
    type: RZ_CMD_DESC_TYPE_ARGV_STATE
```

File: librz/core/cmd/cmd_analysis.c [hunk 2]
```diff
@@ -8812,3 +8812,36 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
 }
+
+RZ_IPI RzCmdStatus rz_analyze_bytes_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	ut8 *buf;
+	int len;
+
+	if (argc < 2) {
+		return RZ_CMD_STATUS_ERROR;
+	}
+
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: no need for this, it is already checked by the parser

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/core/cmd/cmd_analysis.c [hunk 3]
```diff
@@ -8812,3 +8812,36 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+	len = rz_hex_str2bin(argv[1], buf);
+	if (len <= 0) {
+		free(buf);
+		return RZ_CMD_STATUS_ERROR;
+	}
+
+	switch (mode) {
+	case RZ_OUTPUT_MODE_JSON:
+		core_analysis_bytes(core, buf, len, 0, 'j');
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: I recommend to refactor that `core_analysis_bytes()` function to use the `RZ_OUTPUT_MODE_*` instead, also a better name probably. In the same PR, otherwise it might get forgotten.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by PeiweiHu** [comment]
No need to review now. This commit is for the convenience of coding from home. :-)

File: librz/core/cmd_descs/cmd_analysis.yaml [hunk 4]
```diff
@@ -1459,7 +1459,66 @@ commands:
...
+      - RZ_OUTPUT_MODE_JSON
+  - name: aO
+    summary: Analyze N instructions in M bytes
+    subcommands:
+      - name: aO
+        summary: Analyze next N bytes as instructions
+        cname: analyze_n_bytes
+        args:
+          - name: byte_len
+            type: RZ_CMD_ARG_TYPE_STRING
```
Comment (wargio) [inline librz/core/cmd_descs/cmd_analysis.yaml]:
```suggestion
            type: RZ_CMD_ARG_TYPE_NUM
```

File: librz/core/cmd_descs/cmd_analysis.yaml [hunk 5]
```diff
@@ -1459,7 +1459,66 @@ commands:
...
\ No newline at end of file
+      - RZ_OUTPUT_MODE_JSON
+  - name: aO
+    summary: Analyze N instructions in M bytes
+    subcommands:
+      - name: aO
+        summary: Analyze next N bytes as instructions
+        cname: analyze_n_bytes
+        args:
+          - name: byte_len
```
Comment (wargio) [inline librz/core/cmd_descs/cmd_analysis.yaml]: maybe is better to call this `n_bytes`

File: librz/core/cmd/cmd_analysis.c [hunk 6]
```diff
@@ -904,347 +993,378 @@ static void core_analysis_bytes(RzCore *core, const ut8 *buf, int len, int nops,
...
+			op.fail = hint->fail;
+		}
+		if (op.jump != UT64_MAX) {
+			pj_kn(pj, "jump", op.jump);
+		}
+		if (op.fail != UT64_MAX) {
+			pj_kn(pj, "fail", op.fail);
+		}
+		const char *jesil = (hint && hint->esil) ? hint->esil : esilstr;
+		if (jesil && *jesil) {
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: RZ_STR_ISNOTEMPTY

File: librz/core/cmd/cmd_analysis.c [hunk 7]
```diff
@@ -904,347 +993,378 @@ static void core_analysis_bytes(RzCore *core, const ut8 *buf, int len, int nops,
...
+			pj_ks(pj, "esil", jesil);
+		}
+		if (op.il_op) {
+			pj_k(pj, "rzil");
+			rz_il_op_effect_json(op.il_op, pj);
+		}
+		pj_kb(pj, "sign", op.sign);
+		pj_kn(pj, "prefix", op.prefix);
+		pj_ki(pj, "id", op.id);
+		if (opexstr && *opexstr) {
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: RZ_STR_ISNOTEMPTY

File: librz/core/cmd/cmd_analysis.c [hunk 8]
```diff
@@ -904,347 +993,378 @@ static void core_analysis_bytes(RzCore *core, const ut8 *buf, int len, int nops,
...
+		// 0x33->sym.xx
+		char *p = strdup(strsub);
+		if (p) {
+			rz_parse_filter(core->parser, addr, core->flags, hint, p,
+				strsub, sizeof(strsub), be);
+			free(p);
+		}
+		pj_o(pj);
+		pj_ks(pj, "opcode", rz_asm_op_get_asm(&asmop));
+		if (!*strsub) {
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c:994]: RZ_STR_ISEMPTY

File: librz/core/cmd/cmd_analysis.c [hunk 9]
```diff
@@ -8812,3 +8862,236 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+	case RZ_OUTPUT_MODE_STANDARD:
+		core_analysis_bytes_standard(core, core->block, len, 0);
+		break;
+	default:
+		rz_warn_if_reached();
+	}
+
+	if (tbs != core->blocksize) {
+		rz_core_block_size(core, tbs);
+	}
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Avoid using `int`, especially for sizes, since they cannot be negative (in this case `ut32` should be ok). also throw errors for invalid values, like `zero` in this case.

File: librz/core/cmd/cmd_analysis.c [hunk 10]
```diff
@@ -8812,3 +8862,236 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_JSON:
+		core_analysis_bytes_json(core, core->block, len, 0, state->d.pj);
+		break;
+	case RZ_OUTPUT_MODE_STANDARD:
+		core_analysis_bytes_standard(core, core->block, len, 0);
+		break;
+	default:
+		rz_warn_if_reached();
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c:8772]: add a `break` there.

**Comment by PeiweiHu** [comment]
This pr ports command ao, aO, and a8 to new rzshell. The following changes are performed in detail. 

(e->esil, s->size, d->description)

+ split function `core_analysis_bytes()` to `core_analysis_bytes_esil/size/desc/json/standard()`
+ aO(j), aOd, aOs, aOe in rzshell
+ a8(j) in rzshell
+ ao(j), aos, aoe, aod, aoda, aoc, aom, aoma in rzshell (Note that aoef and ao* don't have implementation in old shell. Thus they are removed. )

File: librz/core/cmd/cmd_analysis.c (hunk 8 above)
Comment (PeiweiHu) [reply librz/core/cmd/cmd_analysis.c:994]: Thanks for your reviewing. This leads to `warning: the address of ‘strsub’ will always evaluate as ‘true’ [-Waddress]   64 | #define RZ_STR_ISEMPTY(x)    (!(x) || !*(x))` in compile. Is that ok?

File: librz/core/cmd/cmd_analysis.c [hunk 11]
```diff
@@ -8893,9 +8893,15 @@ RZ_IPI RzCmdStatus rz_analyze_bytes_handler(RzCore *core, int argc, const char *
...
 
 RZ_IPI RzCmdStatus rz_analyze_n_bytes_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
-	int len = core->blocksize;
-	int tbs = len;
-	len = (int)rz_num_get(core->num, argv[1]);
+	ut32 len = core->blocksize;
+	ut32 tbs = len;
+
+	len = rz_num_get(core->num, argv[1]);
+	if ((int)len <= 0) {
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
	if (len < 1) {
```
no need to test it for negative numbers since `rz_num_get` returns unsigned 64.

File: librz/core/cmd/cmd_analysis.c [hunk 12]
```diff
@@ -8893,9 +8893,15 @@ RZ_IPI RzCmdStatus rz_analyze_bytes_handler(RzCore *core, int argc, const char *
...
+		eprintf("Invalid zero or negative arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: RZ_LOG_ERROR here. Use `eprintf` only to debug your code, when you push it to the PR, there should not be any.
use RZ_LOG_* functions for logging errors, warning, infos and debug messages and use rz_cons_* for anything that needs to be printed like you would do using printf

File: librz/core/cmd/cmd_analysis.c [hunk 13]
```diff
@@ -8918,9 +8925,15 @@ RZ_IPI RzCmdStatus rz_analyze_n_bytes_handler(RzCore *core, int argc, const char
...
 RZ_IPI RzCmdStatus rz_analyze_n_bytes_esil_handler(RzCore *core, int argc, const char **argv) {
-	int len = core->blocksize;
-	int tbs = len;
-	len = (int)rz_num_get(core->num, argv[1]);
+	ut32 len = core->blocksize;
+	ut32 tbs = len;
+
+	len = rz_num_get(core->num, argv[1]);
+	if ((int)len <= 0) {
+		eprintf("Invalid zero or negative arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 14]
```diff
@@ -8932,9 +8945,15 @@ RZ_IPI RzCmdStatus rz_analyze_n_bytes_esil_handler(RzCore *core, int argc, const
...
 RZ_IPI RzCmdStatus rz_analyze_n_bytes_desc_handler(RzCore *core, int argc, const char **argv) {
-	int len = core->blocksize;
-	int tbs = len;
-	len = (int)rz_num_get(core->num, argv[1]);
+	ut32 len = core->blocksize;
+	ut32 tbs = len;
+
+	len = rz_num_get(core->num, argv[1]);
+	if ((int)len <= 0) {
+		eprintf("Invalid zero or negative arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 15]
```diff
@@ -8946,9 +8965,15 @@ RZ_IPI RzCmdStatus rz_analyze_n_bytes_desc_handler(RzCore *core, int argc, const
...
 RZ_IPI RzCmdStatus rz_analyze_n_bytes_size_handler(RzCore *core, int argc, const char **argv) {
-	int len = core->blocksize;
-	int tbs = len;
-	len = (int)rz_num_get(core->num, argv[1]);
+	ut32 len = core->blocksize;
+	ut32 tbs = len;
+
+	len = rz_num_get(core->num, argv[1]);
+	if ((int)len <= 0) {
+		eprintf("Invalid zero or negative arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 16]
```diff
@@ -8960,13 +8985,16 @@ RZ_IPI RzCmdStatus rz_analyze_n_bytes_size_handler(RzCore *core, int argc, const
...
-	int l, count = 1, obs = core->blocksize;
+	ut32 l, count = 1, obs = core->blocksize;
 
 	if (argc > 1) {
-		l = (int)rz_num_get(core->num, argv[1]);
-		if (l > 0) {
-			count = l;
+		l = rz_num_get(core->num, argv[1]);
+		if ((int)l <= 0) {
+			eprintf("Invalid zero or negative arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 17]
```diff
@@ -8991,13 +9020,15 @@ RZ_IPI RzCmdStatus rz_analyze_n_ins_handler(RzCore *core, int argc, const char *
...
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 18]
```diff
@@ -9013,13 +9044,15 @@ RZ_IPI RzCmdStatus rz_analyze_n_ins_size_handler(RzCore *core, int argc, const c
...
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 19]
```diff
@@ -9063,14 +9096,18 @@ RZ_IPI RzCmdStatus rz_analyze_cycles_handler(RzCore *core, int argc, const char
...
 	int cr = rz_config_get_i(core->config, "asm.cmt.right");
 	int fun = rz_config_get_i(core->config, "asm.functions");
 	int li = rz_config_get_i(core->config, "asm.lines");
 	int xr = rz_config_get_i(core->config, "asm.xrefs");
 
 	if (argc > 1) {
-		ccl = (int)rz_num_get(core->num, argv[1]);
+		ccl = rz_num_get(core->num, argv[1]);
+		if ((int)ccl < 0) {
+			eprintf("Invalid negative arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 20]
```diff
@@ -9093,5 +9130,60 @@ RZ_IPI RzCmdStatus rz_analyze_cycles_handler(RzCore *core, int argc, const char
...
+}
+
+RZ_IPI RzCmdStatus rz_convert_mne_handler(RzCore *core, int argc, const char **argv) {
+	ut32 id;
+
+	if(rz_str_isnumber(argv[1])) {
+		id = rz_num_math(core->num, argv[1]);
+		// id starts from 1
+		if ((int)id <= 0) {
+			eprintf("Invalid negative or zero arguments.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: Same as explained above

File: librz/core/cmd/cmd_analysis.c [hunk 21]
```diff
@@ -9093,5 +9130,60 @@ RZ_IPI RzCmdStatus rz_analyze_cycles_handler(RzCore *core, int argc, const char
...
+			return RZ_CMD_STATUS_ERROR;
+		}
+
+		char *ops = rz_asm_mnemonics(core->rasm, id, false);
+		if (!ops) {
+			eprintf("Can not find mnemonic by id.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: RZ_LOG_ERROR

File: librz/core/cmd/cmd_analysis.c [hunk 22]
```diff
@@ -9093,5 +9130,60 @@ RZ_IPI RzCmdStatus rz_analyze_cycles_handler(RzCore *core, int argc, const char
...
+			return RZ_CMD_STATUS_ERROR;
+		}
+
+		rz_cons_println(ops);
+		free(ops);
+	} else {
+		id = rz_asm_mnemonics_byname(core->rasm, argv[1]);
+		if (id <= 0) {
+			eprintf("Can not find id by mnemonic.\n");
```
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: RZ_LOG_ERROR

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/core/cmd/cmd_analysis.c (hunk 8 above)
Comment (wargio) [reply librz/core/cmd/cmd_analysis.c:994]: keep it in this way then.

File: librz/core/cmd/cmd_analysis.c (hunk 11 above)
Comment (PeiweiHu) [reply librz/core/cmd/cmd_analysis.c]: Thanks for your reviewing. IMO, although `rz_num_get()` returns an unsigned number, it origins from `argv[1]`, which may be a negative number. For example, when user types `-1` as `argv[1]`, `len` will be `4294967295`. Maybe it's better to cast `len` to `int` temporarily for negative check?

File: librz/core/cmd/cmd_analysis.c (hunk 12 above)
Comment (PeiweiHu) [reply librz/core/cmd/cmd_analysis.c]: Thanks for the detailed explanation. :-)

File: librz/core/cmd/cmd_analysis.c (hunk 11 above)
Comment (wargio) [reply librz/core/cmd/cmd_analysis.c]: then just cast it to `st32` (change also the type from ut32 to st32) and just keep that check. as said before, avoid using `int` and enforce instead bitsizes via `ut<bits>` and `st<bits>` types

**Review by wargio — CHANGES_REQUESTED** [review]
Please add tests also for the a0 and subcommands that you have added.

**Comment by wargio** [comment]
after adding the tests, i think we can merge it! anyway great job <3

File: librz/core/cmd/cmd_analysis.c [hunk 23]
```diff
@@ -8812,3 +8740,329 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+	}
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_analyze_n_ins_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	st32 l;
+	ut32 count = 1, obs = core->blocksize;
+
+	if (argc > 1) {
+		l = (st32)rz_num_get(core->num, argv[1]);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: I'm not sure yet about the real diff between `rz_num_get` and `rz_num_math` but I think we usually used `rz_num_math`in command handlers so far... @thestr4ng3r do you know?

File: librz/core/cmd/cmd_analysis.c [hunk 24]
```diff
@@ -8812,3 +8740,329 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+RZ_IPI RzCmdStatus rz_analyze_cycles_handler(RzCore *core, int argc, const char **argv) {
+	RzList *hooks;
+	RzListIter *iter;
+	RzAnalysisCycleHook *hook;
+	char *instr_tmp = NULL;
+	st32 ccl = 0;
+	int cr = rz_config_get_i(core->config, "asm.cmt.right");
+	int fun = rz_config_get_i(core->config, "asm.functions");
+	int li = rz_config_get_i(core->config, "asm.lines");
+	int xr = rz_config_get_i(core->config, "asm.xrefs");
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: I suggest to use `RzConfigHold`, `rz_config_hold_new` , etc. instead of these.

Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: I suggest to use `RzConfigHold`, `rz_config_hold_new` , etc. instead of these.

File: librz/core/cmd/cmd_analysis.c [hunk 25]
```diff
@@ -8812,3 +8740,329 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+		l = (st32)rz_num_get(core->num, argv[1]);
+		if (l <= 0) {
+			RZ_LOG_ERROR("Invalid zero or negative arguments.\n");
+			return RZ_CMD_STATUS_ERROR;
+		}
+
+		count = l;
+		l *= 8;
+		if (l > obs) {
+			rz_core_block_size(core, l);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:8808]: Do you know why if this command works on "number of instructions" we modify the block size? I think there is something wrong here.

Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:8808]: Do you know why if this command works on "number of instructions" we modify the block size? I think there is something wrong here.

File: librz/core/cmd/cmd_analysis.c [hunk 26]
```diff
@@ -8812,3 +8740,329 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+	if (argc > 1) {
+		l = (st32)rz_num_get(core->num, argv[1]);
+		if (l <= 0) {
+			RZ_LOG_ERROR("Invalid zero or negative arguments.\n");
+			return RZ_CMD_STATUS_ERROR;
+		}
+		count = l;
+		l *= 8;
+		if (l > obs) {
+			rz_core_block_size(core, l);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:8843]: same here. command takes "number of instructions" but here we modify the block size.

Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:8843]: same here. command takes "number of instructions" but here we modify the block size.

File: librz/core/cmd/cmd_analysis.c [hunk 27]
```diff
@@ -8812,3 +8740,329 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:8868]: here as well.

Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:8868]: here as well.

File: librz/core/cmd_descs/cmd_analysis.yaml [hunk 28]
```diff
@@ -1453,3 +1453,100 @@ commands:
...
+    type: RZ_CMD_DESC_TYPE_ARGV_STATE
+    modes:
+      - RZ_OUTPUT_MODE_STANDARD
+      - RZ_OUTPUT_MODE_JSON
+  - name: aO
+    summary: Analyze N instructions in M bytes
+    subcommands:
+      - name: aO
+        summary: Analyze next N bytes as instructions
+        cname: analyze_n_bytes
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml:1471]: this, aOe, aOd and aOs all works on bytes and essentially just change the block size.
I propose to *remove* the n_bytes argument from these command handlers and just rely on core->blocksize by default. Also, change the description of these commands to say "analyze ... next block as instructions".
Users can easily use `aO @!10` to set the block size to 10, so we don't need these command handlers to do the same work.

cc @wargio

Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml:1471]: this, aOe, aOd and aOs all works on bytes and essentially just change the block size.
I propose to *remove* the n_bytes argument from these command handlers and just rely on core->blocksize by default. Also, change the description of these commands to say "analyze ... next block as instructions".
Users can easily use `aO @!10` to set the block size to 10, so we don't need these command handlers to do the same work.

cc @wargio

**Review by ret2libc — CHANGES_REQUESTED** [review]
Thanks for this! It's great to see more commands being moved to rzshell. I left some comments, please let me know if something is not clear.

File: librz/core/cmd_descs/cmd_analysis.yaml (hunk 28 above)
Comment (wargio) [reply librz/core/cmd_descs/cmd_analysis.yaml:1471]: that would simplify the commands by a lot; i think this is a good suggestion, therefore i agree on this.

File: librz/core/cmd/cmd_analysis.c (hunk 25 above)
Comment (PeiweiHu) [reply librz/core/cmd/cmd_analysis.c:8808]: Thanks for your review. IMO, the "number of instructions" part is controlled by `i < nops` in `for (i = idx = 0; idx < len && (!nops || (nops && i < nops)); i++, idx += ret) {`. As for the modification for block size, maybe the original coder assumed every instruction has 8 bytes, thus, there should be at least `count*8` bytes in `core->block` for the following analysis. And `rz_core_block_size()` is called if `core->block` is not big enough.

Comment (ret2libc) [reply librz/core/cmd/cmd_analysis.c:8808]: There is no need to change the blocksize in these cases, I think. The operation should be based on the number of instructions and keep reading until N instructions are analyzed. I guess the old code was already doing this hackish thing :( so... let's keep it this way for now, unless it is easy to change right now.

File: librz/core/cmd/cmd_analysis.c [hunk 29]
```diff
@@ -845,32 +830,121 @@ static void cmd_syscall_do(RzCore *core, st64 n, ut64 addr) {
...
-			return;
+	for (i = idx = 0; idx < len && (!nops || (nops && i < nops)); i++, idx += ret) {
+		addr = core->offset + idx;
+		rz_asm_set_pc(core->rasm, addr);
+		ret = rz_analysis_op(core->analysis, &op, addr, buf + idx, len - idx,
+			RZ_ANALYSIS_OP_MASK_ESIL | RZ_ANALYSIS_OP_MASK_IL | RZ_ANALYSIS_OP_MASK_OPEX | RZ_ANALYSIS_OP_MASK_HINT);
+		esilstr = RZ_STRBUF_SAFEGET(&op.esil);
+
+		if (ret < 1) {
+			eprintf("Oops at 0x%08" PFMT64x " (", core->offset + idx);
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Same.

File: librz/core/cmd/cmd_analysis.c [hunk 30]
```diff
@@ -845,32 +830,121 @@ static void cmd_syscall_do(RzCore *core, st64 n, ut64 addr) {
...
+			for (i = idx, j = 0; i < core->blocksize && j < 3; i++, j++) {
+				eprintf("%02x ", buf[i]);
+			}
+			eprintf("...)\n");
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Same.

File: librz/core/cmd/cmd_analysis.c [hunk 31]
```diff
@@ -845,32 +830,121 @@ static void cmd_syscall_do(RzCore *core, st64 n, ut64 addr) {
...
+	int totalsize = 0;
+
+	for (i = idx = 0; idx < len && (!nops || (nops && i < nops)); i++, idx += ret) {
+		addr = core->offset + idx;
+		rz_asm_set_pc(core->rasm, addr);
+		ret = rz_analysis_op(core->analysis, &op, addr, buf + idx, len - idx,
+			RZ_ANALYSIS_OP_MASK_ESIL | RZ_ANALYSIS_OP_MASK_IL | RZ_ANALYSIS_OP_MASK_OPEX | RZ_ANALYSIS_OP_MASK_HINT);
+
+		if (ret < 1) {
+			eprintf("Oops at 0x%08" PFMT64x " (", core->offset + idx);
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Use `RZ_LOG_ERROR()`. Also, I suggest better error message.

File: librz/core/cmd/cmd_analysis.c [hunk 32]
```diff
@@ -845,32 +830,121 @@ static void cmd_syscall_do(RzCore *core, st64 n, ut64 addr) {
...
+
+		char *opname = strdup(rz_asm_op_get_asm(&asmop));
+		if (opname) {
+			rz_str_split(opname, ' ');
+			char *d = rz_asm_describe(core->rasm, opname);
+			if (d && *d) {
+				rz_cons_printf("%s: %s\n", opname, d);
+				free(d);
+			} else {
+				eprintf("Unknown opcode\n");
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: `RZ_LOG_ERROR()`, also a more detailed error message.

File: librz/core/cmd/cmd_analysis.c [hunk 33]
```diff
@@ -894,7 +968,7 @@ static void core_analysis_bytes(RzCore *core, const ut8 *buf, int len, int nops,
 				mnem = arg;
 			}
 		}
-		if (ret < 1 && fmt != 'd') {
+		if (ret < 1) {
 			eprintf("Oops at 0x%08" PFMT64x " (", core->offset + idx);
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Same

File: librz/core/cmd/cmd_analysis.c [hunk 34]
```diff
@@ -894,7 +968,7 @@ static void core_analysis_bytes(RzCore *core, const ut8 *buf, int len, int nops,
...
 			for (i = idx, j = 0; i < core->blocksize && j < 3; i++, j++) {
 				eprintf("%02x ", buf[i]);
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Same

File: librz/core/cmd/cmd_analysis.c [hunk 35]
```diff
@@ -904,347 +978,378 @@ static void core_analysis_bytes(RzCore *core, const ut8 *buf, int len, int nops,
...
-			if (op.direction != 0) {
-				const char *dir = op.direction == 1 ? "read"
-					: op.direction == 2         ? "write"
-					: op.direction == 4         ? "exec"
-					: op.direction == 8         ? "ref"
-								    : "none";
-				printline("direction", "%s\n", dir);
+		}
+		if (ret < 1) {
+			eprintf("Oops at 0x%08" PFMT64x " (", core->offset + idx);
```
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Same

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd/cmd_analysis.c (hunk 29 above)
Comment (PeiweiHu) [reply librz/core/cmd/cmd_analysis.c]: Thanks for your review. Changed as you commented.

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
Please don't forget to address @ret2libc's comments.

**Comment by PeiweiHu** [comment]
> Please don't forget to address @ret2libc's comments.

It seems there is only a "number of instructions" staff left which @ret2libc said `keep it this way for now`. Any other unsolved comments? With pleasure to accept comments.

**Comment by wargio** [comment]
> > Please don't forget to address @ret2libc's comments.
> 
> It seems there is only a "number of instructions" staff left which @ret2libc said `keep it this way for now`. Any other unsolved comments? With pleasure to accept comments.

mark them as resolverd, then

**Comment by PeiweiHu** [comment]
> > > Please don't forget to address @ret2libc's comments.
> > 
> > 
> > It seems there is only a "number of instructions" staff left which @ret2libc said `keep it this way for now`. Any other unsolved comments? With pleasure to accept comments.
> 
> mark them as resolverd, then

Aw, done. Thanks for the reminder.

File: librz/core/cmd/cmd_analysis.c [hunk 36]
```diff
@@ -8812,3 +8729,274 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+	RzListIter *iter;
+	RzAnalysisCycleHook *hook;
+	char *instr_tmp = NULL;
+	st32 ccl = 0;
+	RzConfigHold *hc = rz_config_hold_new(core->config);
+
+	rz_config_hold_i(hc, "asm.cmt.right", NULL);
+	rz_config_hold_i(hc, "asm.functions", NULL);
+	rz_config_hold_i(hc, "asm.lines", NULL);
+	rz_config_hold_i(hc, "asm.xrefs", NULL);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
	rz_config_hold_i(hc, "asm.cmt.right", "asm.functions", "asm.lines", "asm.xrefs", NULL);
```

not super important, JFYI

File: librz/core/cmd/cmd_analysis.c [hunk 37]
```diff
@@ -8812,3 +8729,274 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+
+RZ_IPI RzCmdStatus rz_analyze_n_bytes_desc_handler(RzCore *core, int argc, const char **argv) {
+	core_analysis_bytes_desc(core, core->block, core->blocksize, 0);
+
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_analyze_n_bytes_size_handler(RzCore *core, int argc, const char **argv) {
+	core_analysis_bytes_size(core, core->block, core->blocksize, 0);
+
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
```

File: librz/core/cmd/cmd_analysis.c [hunk 38]
```diff
@@ -8812,3 +8729,274 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+RZ_IPI RzCmdStatus rz_analyze_n_bytes_esil_handler(RzCore *core, int argc, const char **argv) {
+	core_analysis_bytes_esil(core, core->block, core->blocksize, 0);
+
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_analyze_n_bytes_desc_handler(RzCore *core, int argc, const char **argv) {
+	core_analysis_bytes_desc(core, core->block, core->blocksize, 0);
+
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
```

File: librz/core/cmd/cmd_analysis.c [hunk 39]
```diff
@@ -8812,3 +8729,274 @@ RZ_IPI RzCmdStatus rz_analysis_class_vtable_lookup_handler(RzCore *core, int arg
...
+		rz_warn_if_reached();
+		break;
+	}
+
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_analyze_n_bytes_esil_handler(RzCore *core, int argc, const char **argv) {
+	core_analysis_bytes_esil(core, core->block, core->blocksize, 0);
+
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
```

File: librz/core/cmd/cmd_analysis.c (hunk 36 above)
Comment (PeiweiHu) [reply librz/core/cmd/cmd_analysis.c]: Thanks for your comments. Already changed.

**Review by ret2libc — APPROVED** [review]
